torch/utils/data/datapipes/iter/combining.py

```diff
@@ -0,0 +1,188 @@
+from collections import defaultdict, deque
+from torch.utils.data import IterDataPipe
+from typing import Dict, Iterator, Optional, Sequence, Sized, Tuple, TypeVar
+
+T_co = TypeVar('T_co', covariant=True)
+
+
+class UnzipIterDataPipe(IterDataPipe):
+    r""" :class:`UnzipIterDataPipe`.
+
+    The inverse of :class:`ZipIterDataPipe`, this DataPipe disaggregates
+    the elements of the zipped iterable DataPipe. The first element
+    determines the number of output DataPipe. A :class:`ZipIterDataPipe`
+    is often attached later to aggregate these output DataPipes back.
+    Return a tuple of split DataPipes.
+
+    Example:
+        >>> print(list(dp))
+        [(1, 2), (3, 4), (5, 6), ...]
+        >>> dp1, dp2 = datapipes.iter.Unzip(dp)
+        >>> print(list(dp1))
+        [1, 3, 5, ...]
+        >>> print(list(dp2))
+        [2, 3, 4, ...]
+        >>> def fn(data):
+        ...     return torch.tensor(data)
+        >>> map_dp = datapipes.iter.Map(dp1, fn=fn)
+        >>> zipped_dp = datapipes.iter.Zip(map_dp, dp2)
+        >>> print(list(zipped_dp))
+        [(tensor(1), 2), (tensor(3), 4), (tensor(5), 6), ...]
+
+    .. note:: This DataPipe may require significant storage, especially
+              when one iterator uses most of the data before another
+              iterator starts.
+
+    args:
+        datapipe: Iterable DataPipe being disaggregated
+    """
+    datapipe: IterDataPipe[Sequence]
+    num_splits: int
+    _split_stopped: Dict[int, bool]
+    _all_stopped: bool
+    _it: Iterator[Sequence]
+
+    def __new__(cls, datapipe: IterDataPipe[Sequence]):
+        source_dp = super().__new__(cls)
+        # TODO(Erjia): Valid type inference mechanism without fetching data
+        #              at construction time
+        it = iter(datapipe)
+        data = next(it)
+        num_splits = len(data)
+        source_dp.__init__(datapipe, num_splits, it, data)
```

Comment:
I have to get one data at the construction time to get the number of elements from the source DataPipe.
As @VitalyFedyunin suggested, it will ruin our intention to have lazy_init for the pipeline. In order to get rid of data fetching at construction time, we have to have a mechanism to infer or get the data type for all DataPipes.